dot/sync/chain_processor.go

```diff
@@ -0,0 +1,290 @@
+				// depending on the error, we might want to save this block for later
+			}
+		case <-s.ctx.Done():
+			return
+		}
+	}
+}
+
+// processBlockData processes the BlockData from a BlockResponse and returns the index of the last BlockData it handled on success,
+// or the index of the block data that errored on failure.
+func (s *chainProcessor) processBlockData(bd *types.BlockData) error {
+	if bd == nil {
+		return ErrNilBlockData
+	}
+
+	err := s.blockState.CompareAndSetBlockData(bd)
+	if err != nil {
+		return fmt.Errorf("failed to compare and set data: %w", err)
+	}
+
+	hasHeader, _ := s.blockState.HasHeader(bd.Hash)
+	hasBody, _ := s.blockState.HasBlockBody(bd.Hash)
+	if hasHeader && hasBody {
+		// TODO: fix this; sometimes when the node shuts down the "best block" isn't stored properly,
+		// so when the node restarts it has blocks higher than what it thinks is the best, causing it not to sync
+		// if we update the node to only store finalised blocks in the database, this should be fixed and the entire
+		// code block can be removed
+		block, err := s.blockState.GetBlockByHash(bd.Hash) //nolint
+		if err != nil {
+			logger.Debug("failed to get header", "hash", bd.Hash, "error", err)
+			return err
+		}
+
+		logger.Debug("skipping block, already have", "hash", bd.Hash, "number", block.Header.Number)
+
+		err = s.blockState.AddBlockToBlockTree(block.Header)
+		if err != nil && !errors.Is(err, blocktree.ErrBlockExists) {
+			logger.Warn("failed to add block to blocktree", "hash", bd.Hash, "error", err)
+			return err
+		}
+
+		if errors.Is(err, blocktree.ErrBlockExists) {
+			return nil
+		}
+
+		if bd.Justification != nil && bd.Justification.Exists() {
+			logger.Debug("handling Justification...", "number", block.Header.Number, "hash", bd.Hash)
+			s.handleJustification(block.Header, bd.Justification.Value())
+		}
+
+		// TODO: this is probably unnecessary, since the state is already in the database
+		// however, this case shouldn't be hit often, since it's only hit if the node state
+		// is rewinded or if the node shuts down unexpectedly
+		state, err := s.storageState.TrieState(&block.Header.StateRoot)
+		if err != nil {
+			logger.Warn("failed to load state for block", "block", block.Header.Hash(), "error", err)
+			return err
+		}
+
+		if err := s.blockImportHandler.HandleBlockImport(block, state); err != nil {
+			logger.Warn("failed to handle block import", "error", err)
+		}
+
+		return nil
+	}
+
+	var header *types.Header
+
+	if bd.Header.Exists() && !hasHeader {
+		header, err = types.NewHeaderFromOptional(bd.Header)
+		if err != nil {
+			return err
+		}
+
+		if err = s.handleHeader(header); err != nil {
+			return err
+		}
+	}
+
+	if bd.Body.Exists() && !hasBody {
+		body, err := types.NewBodyFromOptional(bd.Body) //nolint
+		if err != nil {
+			return err
+		}
+
+		if err = s.handleBody(body); err != nil {
+			return err
+		}
+	}
+
+	if bd.Header.Exists() && bd.Body.Exists() {
```

Comment:
You could remove the previus two conditions and use just this condition
```go
if bd.Header.Exists() && bd.Body.Exists() {
  header, err = types.NewHeaderFromOptional(bd.Header)
  ...
  ...s.handleHeader(header)
  ...
  body, err := types.NewBodyFromOptional(bd.Body)
  ...s.handleBody(body)
}
```

Reply:
yeah, it was previously separated because we might get a `BlockData` with only the header, or only the body, but I think now we will always be passing in `BlockData` with both. will update!